### ISSUE TITLE
Write front end authentication that integrates cognito with SSH add on

### ISSUE THREAD

**Issue by angiedear** [oceanbuilders/smartpod#61]
Add on process will prompt user to authenticate with cognito and handle 401 and 403 errors returned from the remote. 

5 points

**Comment by olagalal** [comment]
doing some research ..

**Comment by omerll** [comment]
Still in progress

**Comment by angiedear** [comment]
@olagalal can you please update us on this issue?

**Comment by scottdear** [comment]
Spoke w/ Ola in discord:

https://docs.diffbot.com/docs/en/guides-injecting-javascript-replaying-ajax

Injecting JavaScript into Custom API and replaying AJAX calls · Doc...

Basically create xhr, if it fails w/ a 401, add it to an array, then redirect them to the login page via SSO.  When they come back loop through the array and replay the xhr requests.  If it's a 403 display an Unauthorized message.  You probably want to use a wrapper function to send the requests and queue them if they fail w/ a 401, then another to loop through and replay if needed.

**Comment by angiedear** [comment]
Split issue into front end/back end Ola and Talha will work on issues together.

**Comment by angiedear** [comment]
Talha will reach out to Ola today

**Comment by omerll** [comment]
Still in porgress, Arrange a meeting with @siddiqss @olagalal @scottdear @amit

**Comment by angiedear** [comment]
Look at the source code and see how they do it in Home Assistant with Nabu Casa- per Scott

**Comment by godslave05** [comment]
In progress

**Comment by siddiqss** [comment]
In progress

**Comment by omerll** [comment]
@olagalal @godslave05 PLEASE HELP TALHA! He needs support!

**Comment by siddiqss** [comment]
In progress. Need to discuss with the team.

**Comment by omerll** [comment]
Team have questions. they are pinging Scott.... they are blocked until they have an answer

**Comment by siddiqss** [comment]
Added a custom panel for user authentication with Cognito (like Nabu Casa) in HA dev server. Writing JavaScript code for custom panel to exchange access token for auth code which is retrieved when a user is authenticated.

**Comment by siddiqss** [comment]
In progress.

**Comment by omerll** [comment]
In progress, Integration with cognito was implemented
Blocked on Omer, need to omer to install SSL certificates on the dev server for asad
A issue was wrriten, and he will install the certificate ASAP

**Comment by siddiqss** [comment]
Created a custom panel for remote access on HA dev server. When user clicks on authenticate, a new window Cognito window opens and user signs in. After authentication is complete, user is redirected to HA again with Auth code. The JS code in the custom panel reads the Auth code and sends a POST request to Cognito to get auth tokens. Auth tokens are used to verify user. Now user can turn ON and OFF the SSH addon using a button.

Need to install SSH addon on dev server (Ola installed it on prod server).
Waiting for Omer to install SSL certificate on dev server.

**Comment by omerll** [comment]
Is almost done, fixing some bugs and making sure it works stable with HA server over time.

**Comment by siddiqss** [comment]
Fixed bugs. The SSH add-on can be controlled (enabled/disabled) with user authentication. Works for Sea Pod users without admin access. 

Link for code and readme is given below.
https://github.com/oceanbuilders/ha-integrations/tree/master/Remote-Panel

**Comment by siddiqss** [comment]
SSH add-on crashed on dev instance. @omerll  has to re-install the HASS OS on dev server

**Comment by omerll** [comment]
Needs to be checked on the mobile app! doest work well there, because it open a new browser windows and creates the cookie on the web browser, but doesnt create the cookie inside the mobile app, and login fails.

It just work on web

Needs to modify the home assistant app to make it work correctly or modify the client side on the web.

**Comment by amitpujar** [comment]
Progress made for Web. Needs to be tested on mobile devices - dependency on Asad after he tests the add-on. Talha to work with Asad offline to resolve this and drive to closure.  Discuss on Tuesday 6 Dec.

**Comment by omerll** [comment]
Waiting on @godslave05, work in progress, waiting to see Josiahs solution for taking a decision

**Comment by siddiqss** [comment]
On hold. Will create a new issue.
